app/gilded-rose.ts

```diff
@@ -1,69 +1,89 @@
 export class Item {
-  name: string;
-  sellIn: number;
-  quality: number;
-
-  constructor(name, sellIn, quality) {
-    this.name = name;
-    this.sellIn = sellIn;
-    this.quality = quality;
-  }
+    name: string;
+    sellIn: number;
+    quality: number;
+
+    constructor(name, sellIn, quality) {
+        this.name = name;
+        this.sellIn = sellIn;
+        this.quality = quality;
+    }
 }
 
 export class GildedRose {
-  items: Array<Item>;
-
-  constructor(items = [] as Array<Item>) {
-    this.items = items;
-  }
-
-  updateQuality() {
-    for (let i = 0; i < this.items.length; i++) {
-      if (this.items[i].name != 'Aged Brie' && this.items[i].name != 'Backstage passes to a TAFKAL80ETC concert') {
-        if (this.items[i].quality > 0) {
-          if (this.items[i].name != 'Sulfuras, Hand of Ragnaros') {
-            this.items[i].quality = this.items[i].quality - 1
-          }
+    items: Array<Item>;
+
+    constructor(items = [] as Array<Item>) {
+        this.items = items;
+    }
+
+    increaseQuality(item, increment = 1) {
+        item.quality = Math.min(item.quality + increment, 50)
+    }
+
+    decreaseQuality(item, decrement = 1) {
+        item.quality = Math.max(0, item.quality - decrement)
+    }
+
+    dealWithAgedBrie(agedBrie) {
+        if (agedBrie.sellIn >= 0) {
+            this.increaseQuality(agedBrie)
+        } else {
+            this.increaseQuality(agedBrie, 2)
+        }
+    }
+
+    dealWithBackstagePasses(backstagePasses) {
+        if (backstagePasses.sellIn > 10) {
+            this.increaseQuality(backstagePasses);
+        } else if (backstagePasses.sellIn > 5 && backstagePasses.sellIn <= 10) {
+            this.increaseQuality(backstagePasses, 2);
+        } else if (backstagePasses.sellIn > 0 && backstagePasses.sellIn <= 5) {
+            this.increaseQuality(backstagePasses, 3);
+        } else {
+            backstagePasses.quality = 0;
         }
-      } else {
-        if (this.items[i].quality < 50) {
-          this.items[i].quality = this.items[i].quality + 1
-          if (this.items[i].name == 'Backstage passes to a TAFKAL80ETC concert') {
-            if (this.items[i].sellIn < 11) {
-              if (this.items[i].quality < 50) {
-                this.items[i].quality = this.items[i].quality + 1
-              }
+    }
+
+    dealWithConjuredItem(conjuredItem) {
+        if (conjuredItem.sellIn >= 0) {
+            this.decreaseQuality(conjuredItem, 2)
+        } else if (conjuredItem.sellIn < 0) {
+            this.decreaseQuality(conjuredItem, 4)
+        }
+    }
+
+    updateQuality() {
```

Comment:
We're using `this.items[i].name` multiple times in this function, can we extract that into a variable and re-use that variable?